scripts/emulabconfig.py

```diff
@@ -71,69 +127,104 @@ def get_remote_scripts_path(self):
     def get_remote_obj_path(self):
         return os.path.join(self.get_remote_wd(), obj_dir)
 
-    def send_code(self, server):
+    def send_code(self, server, parallel=False):
         log("Sending code to %s" % server)
+        if parallel:
+            for host in self.hosts:
+                self.send_code(host[0])
+            return
         subprocess.check_call("rsync -ave ssh --exclude 'logs/*' " +
                               "--exclude 'docs/*' " +
                               "./ %s:%s/ > /dev/null" % (server,
                                                          self.get_remote_wd()),
                               shell=True, stdout=sys.stdout)
- 
-    def compile_code(self, server, clean=False):
+    
+    def compile_code(self, server, clean=False, parallel=False):
         log("Compiling code on %s" % server)
         clean_cmd = ''
         if clean:
             clean_cmd = 'make clean;'
-        ssh(server,
-            '(cd %s; (%s make -j 8)  > ' % (self.get_remote_wd(), clean_cmd) +
-             '%s/build.log)' % self.get_remote_wd())
-
-    def create_log_dir(self, server):
+        if parallel:
+            pdsh('(cd %s; (%s make -j 8)  > ' % (self.get_remote_wd(), clean_cmd) +
+                         '%s/build.log)' % self.get_remote_wd())
+        else:
+            ssh(server,
+                '(cd %s; (%s make -j 8)  > ' % (self.get_remote_wd(), clean_cmd) +
+                 '%s/build.log)' % self.get_remote_wd())
```

Comment:
If these strings are identical, it's probably better to use the if statement to choose a function, and then invoke it once with just the single complicated string.